tests/testsuite/features_namespaced.rs

```diff
@@ -0,0 +1,999 @@
+//! Tests for namespaced features.
+
+use cargo_test_support::project;
+use cargo_test_support::registry::{Dependency, Package};
+
+#[cargo_test]
+fn gated() {
+    // Need namespaced-features to use `crate:` syntax.
+    Package::new("bar", "1.0.0").publish();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+
+                [dependencies]
+                bar = { version = "1.0", optional = true }
+
+                [features]
+                foo = ["crate:bar"]
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("check")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] failed to parse manifest at `[..]/foo/Cargo.toml`
+
+Caused by:
+  namespaced features with the `crate:` prefix are only allowed on the nightly channel \
+  and requires the `-Z namespaced-features` flag on the command-line
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn dependency_gate_ignored() {
+    // Dependencies with `crate:` features are ignored in the registry if not on nightly.
+    Package::new("baz", "1.0.0").publish();
+    Package::new("bar", "1.0.0")
+        .add_dep(Dependency::new("baz", "1.0").optional(true))
+        .feature("feat", &["crate:baz"])
+        .publish();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+
+                [dependencies]
+                bar = "1.0"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("check")
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr(
+            "\
+[UPDATING] [..]
+[ERROR] no matching package named `bar` found
+location searched: registry `https://github.com/rust-lang/crates.io-index`
+required by package `foo v0.1.0 ([..]/foo)`
+",
+        )
+        .run();
+
+    // Publish a version without namespaced features, it should ignore 1.0.0
+    // an use this instead.
+    Package::new("bar", "1.0.1")
+        .add_dep(Dependency::new("baz", "1.0").optional(true))
+        .feature("feat", &["baz"])
+        .publish();
+    p.cargo("check")
+        .masquerade_as_nightly_cargo()
+        .with_stderr(
+            "\
+[UPDATING] [..]
+[DOWNLOADING] crates ...
+[DOWNLOADED] bar [..]
+[CHECKING] bar v1.0.1
+[CHECKING] foo v0.1.0 [..]
+[FINISHED] [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn dependency_with_crate_syntax() {
+    // Registry dependency uses crate: syntax.
+    Package::new("baz", "1.0.0").publish();
+    Package::new("bar", "1.0.0")
+        .add_dep(Dependency::new("baz", "1.0").optional(true))
+        .feature("feat", &["crate:baz"])
```

Comment:
Yep.  I'm actually not sure how to roll this out, since we'll need to be pretty confident on the design since it affects the index.  At what point do we decide to turn it on?